### PR TITLE
[ROCm] add the bias all row -inf support for jax unfused-attn

### PR DESCRIPTION
**Pull request by wangye805**
# Description

CK fixed the corner case with all row -inf in bias in https://github.com/ROCm/composable_kernel/pull/3326. This also aligns with NV upstream. We added the corresponding jax pytest by modifying the jax unfused-attn.

Fixes # (issue)

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

## Changes
This concludes the https://github.com/ROCm/frameworks-internal/issues/14668 and https://ontrack-internal.amd.com/browse/SWDEV-561757

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: tests/jax/test_fused_attn.py [hunk 1]
```diff
@@ -97,6 +97,12 @@ def general_dot_product_attention(
         logits = logits.reshape((b, h_kv * num_groups, s_q, s_kv))
         # apply post-scale bias
         logits = logits + bias
+        # [ROCm] Detect query rows where ALL bias values are -inf (fully masked out).
+        # These rows would produce NaN in softmax; zero them so softmax yields 0 instead.
+        # Use equality check against -inf so real-valued bias (e.g. 1HSS) is unaffected.
+        if is_hip_extension():
+            bias_all_neg_mask = jnp.all(bias == -jnp.inf, axis=-1, keepdims=True)
+            logits = jnp.where(bias_all_neg_mask, 0, logits)
```
Comment (Micky774) [inline tests/jax/test_fused_attn.py:106]: Correct me if I'm wrong here but zeroing out logits does not produce a zero output for softmax. Instead, a softmax with all logits being zero is simply a uniform distribution. This is fine, since later we manually zero it, but I think the comment needs an update.

File: tests/jax/test_fused_attn.py [hunk 2]
```diff
@@ -125,6 +131,14 @@ def general_dot_product_attention(
         case _:
             raise NotImplementedError(f"Unknown {softmax_type=}")
 
+    # [ROCm] Zero out softmax for fully-masked rows to prevent NaN propagation in backward
+    # Only triggers for -inf mask bias, not real-valued bias (e.g. 1HSS)
+    if bias is not None and is_hip_extension():
+        bias_all_neg_mask = jnp.all(bias == -jnp.inf, axis=-1, keepdims=True)
+        # Expand to match softmax_out 5D shape (b, h_kv, num_groups, s_q, s_kv)
+        bias_all_neg_mask = jnp.expand_dims(bias_all_neg_mask, axis=2)
+        softmax_out = jnp.where(bias_all_neg_mask, 0, softmax_out)
```
Comment (Micky774) [inline tests/jax/test_fused_attn.py]: Can we re-use `bias_all_neg_mask` from earlier to avoid potential drift?

File: tests/jax/test_fused_attn.py [hunk 3]
```diff
@@ -1026,6 +1040,13 @@ def grad_func(func, *args, cp_reverse_out=False, **kwargs):
         if self.dropout_prob > 0.0:
             return
 
+        # [ROCm] Verify no NaN or Inf in forward outputs
+        if is_hip_extension():
+            assert not jnp.any(jnp.isnan(primitive_out)), "Fused attention output contains NaN"
+            assert not jnp.any(jnp.isinf(primitive_out)), "Fused attention output contains Inf"
+            assert not jnp.any(jnp.isnan(reference_out)), "Reference attention output contains NaN"
+            assert not jnp.any(jnp.isinf(reference_out)), "Reference attention output contains Inf"
```
Comment (Micky774) [inline tests/jax/test_fused_attn.py]: Perhaps we can make this a bit more pythonic?
```
for name, out in [("Fused", primitive_out), ("Reference", reference_out)]:                                                                
      assert not jnp.any(jnp.isnan(out)), f"{name} attention output contains NaN"                                                           
      assert not jnp.any(jnp.isinf(out)), f"{name} attention output contains Inf"                                                           
```

File: tests/jax/test_fused_attn.py [hunk 4]
```diff
@@ -1053,6 +1074,18 @@ def check_dqkv(primitive, reference, pad, idx):
...
+        if is_hip_extension():
+            for name, p_grad, r_grad in [
+                ("dq", primitive_dq, reference_dq),
+                ("dk", primitive_dk, reference_dk),
+                ("dv", primitive_dv, reference_dv),
+            ]:
+                assert not jnp.any(jnp.isnan(p_grad)), f"Fused attention {name} contains NaN"
+                assert not jnp.any(jnp.isinf(p_grad)), f"Fused attention {name} contains Inf"
+                assert not jnp.any(jnp.isnan(r_grad)), f"Reference attention {name} contains NaN"
+                assert not jnp.any(jnp.isinf(r_grad)), f"Reference attention {name} contains Inf"
```
Comment (Micky774) [inline tests/jax/test_fused_attn.py]: Similar to above

**Comment by Micky774** [comment]
Looks like CI is failing for incorrect broadcasting shapes

File: tests/jax/test_fused_attn.py (hunk 1 above)
Comment (wangye805) [reply tests/jax/test_fused_attn.py:106]: Updated the comment

File: tests/jax/test_fused_attn.py (hunk 2 above)
Comment (wangye805) [reply tests/jax/test_fused_attn.py]: Adopted the previously generated bias_all_neg_mask and fix the shape broadcasting for other bias shapes (other than B1SS)

File: tests/jax/test_fused_attn.py (hunk 3 above)
Comment (wangye805) [reply tests/jax/test_fused_attn.py]: Updated

File: tests/jax/test_fused_attn.py (hunk 4 above)
Comment (wangye805) [reply tests/jax/test_fused_attn.py]: Updated

File: tests/jax/test_fused_attn.py [hunk 5]
```diff
@@ -1474,6 +1507,95 @@ def test_backward(
...
+            2, 1024, 1024, 12, 12, 64, 64, jnp.bfloat16, QKVLayout.BSHD_BSHD_BSHD,
+            BiasShape._BHSS, id="BHSS-SELF",
+        ),
+        pytest.param(
+            2, 1024, 1024, 12, 12, 64, 64, jnp.bfloat16, QKVLayout.BSHD_BSHD_BSHD,
+            BiasShape._11SS, id="11SS-SELF",
+        ),
+    ],
+)
+def test_backward_bias_all_neg_inf(
```
Comment (ipanfilo) [inline tests/jax/test_fused_attn.py:1527]: I think it should be limited to CK backend then

File: tests/jax/test_fused_attn.py [hunk 6]
```diff
@@ -97,6 +97,13 @@ def general_dot_product_attention(
         logits = logits.reshape((b, h_kv * num_groups, s_q, s_kv))
         # apply post-scale bias
         logits = logits + bias
+        # [ROCm] Detect query rows where ALL bias values are -inf (fully masked out).
+        # These rows would produce NaN in softmax; zero logits to prevent NaN since
+        # the softmax output for these rows is zeroed out below anyway.
+        # Use equality check against -inf so real-valued bias (e.g. 1HSS) is unaffected.
+        if is_hip_extension():
+            bias_all_neg_mask = jnp.all(bias == -jnp.inf, axis=-1, keepdims=True)
+            logits = jnp.where(bias_all_neg_mask, 0, logits)
```
Comment (Micky774) [inline tests/jax/test_fused_attn.py:106]: btw if we're zeroing out the softmax afterwards anyways, why do we need to zero out the logits here? It resolves to nan but doesn't break anything right?

Comment (wangye805) [reply tests/jax/test_fused_attn.py:106]: If we do not zero out the logits, then softmax(logits with all row -inf)=nan, then this nan will be back propagated into the bwd pass, then it will pollute all dq/dk/dv

Comment (Micky774) [reply tests/jax/test_fused_attn.py:106]: But aren't those nans specifically zeroed out in the next block anyways? There's no harm in the nans being produced if they're immediately zeroed out right?

Comment (wangye805) [reply tests/jax/test_fused_attn.py:106]: Those nans in the fwd will back propagate to the dq/dk/dv in this exact attn block, not the next block. Let me ask claude to give you an example soon

Comment (Micky774) [reply tests/jax/test_fused_attn.py:106]: Ah nevermind, I understand now. My confusion centered around a misunderstanding regarding how bwd context was being generated (here we rely on VJP). Looks good, thanks.

Comment (wangye805) [reply tests/jax/test_fused_attn.py:106]: @Micky774 
```
"""
Demonstrates that NaN from softmax([-inf,...,-inf]) back-propagates into dq/dk
even if the softmax output is zeroed out afterwards.

logits = q @ k^T + bias    (bias is all -inf for masked rows)
softmax_out = softmax(logits)
output = softmax_out @ v

Usage:
    python nan_backprop_demo.py

Expected output:
    === Without pre-zeroing logits ===
    Forward output: 0.0
    dq has NaN: True
    dk has NaN: True
    dv has NaN: False

    === With pre-zeroing logits ===
    Forward output: 0.0
    dq has NaN: False
    dk has NaN: False
    dv has NaN: False
"""
import jax
import jax.numpy as jnp


def attn_without_pre_zero(q, k, v, bias):
    """Only zero softmax output after NaN — dq/dk get NaN."""
    logits = q @ k.T + bias               # all -inf for masked rows
    s = jax.nn.softmax(logits)             # softmax([-inf,-inf]) = [nan, nan]
    s = jnp.where(jnp.isnan(s), 0, s)     # zero out NaN in forward
    return jnp.sum(s @ v)


def attn_with_pre_zero(q, k, v, bias):
    """Zero logits before softmax — dq/dk are clean."""
    logits = q @ k.T + bias
    all_neg_inf = jnp.all(bias == -jnp.inf, axis=-1, keepdims=True)
    logits = jnp.where(all_neg_inf, 0, logits)
    s = jax.nn.softmax(logits)
    s = jnp.where(all_neg_inf, 0, s)
    return jnp.sum(s @ v)


# 1 query, 2 keys — bias is all -inf (fully masked row)
q = jax.random.normal(jax.random.PRNGKey(0), (1, 4), dtype=jnp.float32)
k = jax.random.normal(jax.random.PRNGKey(1), (2, 4), dtype=jnp.float32)
v = jax.random.normal(jax.random.PRNGKey(2), (2, 4), dtype=jnp.float32)
bias = jnp.array([[-jnp.inf, -jnp.inf]], dtype=jnp.float32)  # all-inf row

print("=== Without pre-zeroing logits ===")
val, (dq, dk, dv) = jax.value_and_grad(attn_without_pre_zero, argnums=(0, 1, 2))(q, k, v, bias)
print(f"Forward output: {val}")
print(f"dq: {dq}")
print(f"dk: {dk}")
print(f"dv: {dv}")
print(f"dq has NaN: {bool(jnp.any(jnp.isnan(dq)))}")
print(f"dk has NaN: {bool(jnp.any(jnp.isnan(dk)))}")
print(f"dv has NaN: {bool(jnp.any(jnp.isnan(dv)))}")

print()

print("=== With pre-zeroing logits ===")
val, (dq, dk, dv) = jax.value_and_grad(attn_with_pre_zero, argnums=(0, 1, 2))(q, k, v, bias)
print(f"Forward output: {val}")
print(f"dq: {dq}")
print(f"dk: {dk}")
print(f"dv: {dv}")
print(f"dq has NaN: {bool(jnp.any(jnp.isnan(dq)))}")
print(f"dk has NaN: {bool(jnp.any(jnp.isnan(dk)))}")
print(f"dv has NaN: {bool(jnp.any(jnp.isnan(dv)))}")
```

**Comment by Micky774** [comment]
CI failure is unrelated and we have a pass on mi35x

File: tests/jax/test_fused_attn.py (hunk 5 above)
Comment (wangye805) [reply tests/jax/test_fused_attn.py:1527]: Thanks. Added a ck backend guard

File: tests/jax/test_fused_attn.py [hunk 7]
```diff
@@ -1474,6 +1507,98 @@ def test_backward(
...
+        pytest.param(
+            2, 1024, 1024, 12, 12, 64, 64, jnp.bfloat16, QKVLayout.BSHD_BSHD_BSHD,
+            BiasShape._BHSS, id="BHSS-SELF",
+        ),
+        pytest.param(
+            2, 1024, 1024, 12, 12, 64, 64, jnp.bfloat16, QKVLayout.BSHD_BSHD_BSHD,
+            BiasShape._11SS, id="11SS-SELF",
+        ),
+    ],
+)
```
Comment (Micky774) [inline tests/jax/test_fused_attn.py:1522]:
```suggestion
@pytest.mark.parametrize(
    "s_kv, h_kv, bias_shape",
    [
        pytest.param(1024, 12, BiasShape._B1SS, id="B1SS-SELF"),
        pytest.param(512,  12, BiasShape._B1SS, id="B1SS-CROSS"),
        pytest.param(1024, 6,  BiasShape._1HSS, id="1HSS-GQA"),
        pytest.param(1024, 12, BiasShape._BHSS, id="BHSS-SELF"),
        pytest.param(1024, 12, BiasShape._11SS, id="11SS-SELF"),
    ],
)
@pytest.mark.parametrize(
    "b, s_q, h_q, d_qk, d_v, dtype, qkv_layout",
    [(2, 1024, 12, 64, 64, jnp.bfloat16, QKVLayout.BSHD_BSHD_BSHD)],
)
```

Optional suggestion -- if you think we might want to affect `b, s_q, h_q, d_qk, d_v, dtype, qkv_layout` in the future then feel free to ignore this and resolve the comment.

**Review by Micky774 — APPROVED** [review]
Left an optional suggestion, otherwise LGTM.

File: tests/jax/test_fused_attn.py (hunk 7 above)
Comment (wangye805) [reply tests/jax/test_fused_attn.py:1522]: Good suggestions. I committed it

**Review by ipanfilo — APPROVED** [review]
